index.ts

```diff
@@ -0,0 +1,104 @@
+const NativeClient = require('bindings')('hawk_tracer_client').HawkTracerClient;   // loads build/Release/hawk_tracer_client.node
+
+export interface HawkTracerClientOptions {
+    source: string;
+    map_files?: string;
+}
+
+// HT_Event
+interface BareEvent {
+    klass_id: number;
+    timestamp: number;
+    id: number;
+}
+
+// HT_EventKlassInfoEvent
+interface EventKlassInfoEvent extends BareEvent {
+    info_klass_id: number;
+    event_klass_name: string;
+    field_count: number;
+}
+
+function isEventKlassInfoEvent(event: BareEvent): event is EventKlassInfoEvent {
+    const e = event as EventKlassInfoEvent;
+    return e.info_klass_id !== undefined && e.event_klass_name !== undefined;
+}
+
+export interface Event extends BareEvent {
+    klass_name: string;
+}
+
+function isEvent(event: Event): event is Event {
+    return event.klass_name !== undefined;
+}
+
+// HT_CallstackBaseEvent | HT_CallstackIntEvent | HT_CallstackStringEvent
+export interface CallstackEvent extends Event {
+    duration: number;
+    thread_id: number;
+    label?: string | number;
+}
+
+export function isCallstackEvent(event: CallstackEvent): event is CallstackEvent {
+    const e = event as CallstackEvent;
+    return e.duration !== undefined && e.thread_id !== undefined;
+}
+
+export class HawkTracerClient {
+    private _client: any;
+    private _klass_names: (string | undefined)[] = [];
+
+    constructor(option: HawkTracerClientOptions | string) {
+        if (typeof option === "string") {
+            this._client = new NativeClient(option);
+        }
+        else {
+            this._client = new NativeClient(option.source, option.map_files);
+        }
+    }
+
+    public start(): boolean {
+        return this._client.start();
+    }
+
+    public stop(): void {
+        this._client.stop();
+    }
+
+    public onEvents(callback: (events: Event[]) => void): void {
+        this._client.onEvents((events: BareEvent[]) => {
+            this.registerEventKlass(events);
+            callback(this.decorateEventWithKlass(events));
+        });
+    }
+
+    private decorateEventWithKlass(bareEvents: BareEvent[]): Event[] {
+        const events = bareEvents as Event[];
+        events.forEach(e => {
+            // @ts-ignore
+            e.klass_name = this._klass_names[e.klass_id];
```

Comment:
What will happen if event already has klass_name field? Will that be overwritten?

Reply:
It will be overwritten.
I think it's a trade-off between data-safety and usability.
As I mentioned in another comment, for javascript/typescript users, it feels unnatural and uncomfortable to get the type information from somewhere else other than the object itself.
I considered having a generic function that returns the event only when the klass matches, but it would take some time to write, at least for me.
Plus, the term "klass" is being used several times in predefined events in hawktracer. I don't see it being used frequently here.
So my decision came from that it'd be really useful, and that in only rare cases user data would be overwritten.
How about checking whether the event already has "klass_name" field and avoid overwriting it, and just log a warning in such a case?

Reply:
I think we might consider making the C++ layer as thin as possible, expose C++ interfaces almost 'as they are' to JS, and then on top of the low-level JS binding layer, build more user-friendly JS layer, so in case somebody needs access to low-level bindings, can have it.